File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.cpp
class Solution {
public:

    void dfs(int start, vector<vector<int>>& ourVec, vector<bool>& visited){
        
        visited[start] = true;
        for (int i = 0; i < ourVec[start].size(); i++){
            if (!visited[ourVec[start][i]]){
                dfs(ourVec[start][i],ourVec,visited);
            }
        }
        
    }


    int countComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> ourVec(n);
        for (int i = 0; i < edges.size(); i++){
            ourVec[edges[i][0]].push_back(edges[i][1]);
            ourVec[edges[i][1]].push_back(edges[i][0]);
        }
        
        vector<bool> visited(n);
        
        int count = 0;
        
        for (int i = 0; i < n; i++){
            if (!visited[i]){
                count++;
                dfs(i,ourVec,visited);
            }
        }
        
        return count;
        
    }
};

        //source node can be any nodes 
        //0 .. n 
        
        //0 -> dfs(0) -> we will be able to find connected components from 0 
        //1 -> markes as visited from our array (visited[1] = true) 
        

File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.java
class Solution {
    public void dfs(List<List<Integer>> adj, int index, boolean[] visited){
        visited[index] = true; 
        for(int next : adj.get(index)){
            if(visited[next] == false){
                dfs(adj,next,visited); 
            }
        }
    }
    public int countComponents(int n, int[][] edges) {
        List<List<Integer>> adj = new ArrayList<>(); 
        for(int i = 0; i < n; i++){
            adj.add(new ArrayList<>()); 
        }

        for(int[] v : edges){
            int from = v[0], to = v[1]; 
            adj.get(from).add(to); 
            adj.get(to).add(from); 
        }

        boolean[] visited = new boolean[n]; 
        int count = 0; 
        for(int i = 0; i < n; i++){
            if(visited[i] == false){
                dfs(adj,i,visited); 
                count++; 
            }
        }
        return count; 
    }
}